File: 772-construct-quad-tree/construct-quad-tree.py
"""
# Definition for a QuadTree node.
class Node:
    def __init__(self, val, isLeaf, topLeft, topRight, bottomLeft, bottomRight):
        self.val = val
        self.isLeaf = isLeaf
        self.topLeft = topLeft
        self.topRight = topRight
        self.bottomLeft = bottomLeft
        self.bottomRight = bottomRight
"""

class Solution:
    def construct(self, grid: List[List[int]]) -> 'Node':
        """
        :type grid: List[List[int]]
        :rtype: Node
        """
        def buildQuadTree(size, grid, startRow, startCol):
            if size == 1:
                return Node(bool(grid[startRow][startCol]), True)

            countZero = 0
            countOne = 0
            for row in range(startRow, startRow + size):
                for col in range(startCol, startCol + size):
                    if grid[row][col] == 1:
                        countOne += 1
                    else:
                        countZero += 1

            if countOne == size * size:
                return Node(True, True)
            elif countZero == size * size:
                return Node(False, True)
            else:
                halfSize = size // 2
                topLeftNode = buildQuadTree(halfSize, grid, startRow, startCol)
                topRightNode = buildQuadTree(halfSize, grid, startRow, startCol + halfSize)
                bottomLeftNode = buildQuadTree(halfSize, grid, startRow + halfSize, startCol)
                bottomRightNode = buildQuadTree(halfSize, grid, startRow + halfSize, startCol + halfSize)
                return Node(True, False, topLeftNode, topRightNode, bottomLeftNode, bottomRightNode)

        n = len(grid)
        if n == 0:
            return None

        return buildQuadTree(n, grid, 0, 0)

File: 772-construct-quad-tree/construct-quad-tree.java
/*
// Definition for a QuadTree node.
class Node {
    public boolean val;
    public boolean isLeaf;
    public Node topLeft;
    public Node topRight;
    public Node bottomLeft;
    public Node bottomRight;

    
    public Node() {
        this.val = false;
        this.isLeaf = false;
        this.topLeft = null;
        this.topRight = null;
        this.bottomLeft = null;
        this.bottomRight = null;
    }
    
    public Node(boolean val, boolean isLeaf) {
        this.val = val;
        this.isLeaf = isLeaf;
        this.topLeft = null;
        this.topRight = null;
        this.bottomLeft = null;
        this.bottomRight = null;
    }
    
    public Node(boolean val, boolean isLeaf, Node topLeft, Node topRight, Node bottomLeft, Node bottomRight) {
        this.val = val;
        this.isLeaf = isLeaf;
        this.topLeft = topLeft;
        this.topRight = topRight;
        this.bottomLeft = bottomLeft;
        this.bottomRight = bottomRight;
    }
}
*/

class Solution {
    public Node construct(int[][] grid) {
        int n = grid.length;
        return buildQuadTree(grid, 0, 0, n-1, n-1);
    }
    
    private Node buildQuadTree(int[][] grid, int left, int top, int right, int bottom) {
        // Check if the sub-grid has the same value
        boolean isLeaf = true;
        int val = grid[top][left];
        for (int i = top; i <= bottom; i++) {
            for (int j = left; j <= right; j++) {
                if (grid[i][j] != val) {
                    isLeaf = false;
                    break;
                }
            }
            if (!isLeaf) {
                break;
            }
        }
        // If the sub-grid has the same value, create a leaf node
        if (isLeaf) {
            return new Node(val == 1, true, null, null, null, null);
        }
        // If the sub-grid has different values, divide it into four sub-grids
        int midRow = (top + bottom) / 2;
        int midCol = (left + right) / 2;
        Node topLeft = buildQuadTree(grid, left, top, midCol, midRow);
        Node topRight = buildQuadTree(grid, midCol+1, top, right, midRow);
        Node bottomLeft = buildQuadTree(grid, left, midRow+1, midCol, bottom);
        Node bottomRight = buildQuadTree(grid, midCol+1, midRow+1, right, bottom);
        // Create a non-leaf node with the four children
        return new Node(false, false, topLeft, topRight, bottomLeft, bottomRight);
    }
}